### DISCUSSION TITLE
Functions in nested data - context issue?

### DISCUSSION THREAD

**Discussion by jonas-grobe** [alpinejs/alpine#3989] [7. Bugs (archived)] [answered]
I have encountered a possible issue with functions in nested data.

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Alpine test</title>
    <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.3/dist/cdn.js"></script>
</head>

<script>
    document.addEventListener('alpine:init', () => {
        Alpine.data('app', () => ({
            v1: 0,
            f1() { return this.v1; },
            f2() { return this.f1(); },
            inner: {
                v2: 0,
                f3() { return this.inner.v2; },
                f4() { return this.inner.f3(); },
            }
        }))
    });
</script>
<body x-data="app">
    <span x-text="f2"></span>
    <span x-text="inner.f4"></span> <!-- f4 calls f3. f3 fails: this.inner is undefined-->
</body>

</html>
```

I would expect `inner.f4` to behave just like `f2` and return `0`. Instead it correctly calls f3 but f3 fails with `this.inner is undefined`

**Comment by ekwoka** [comment, marked as answer]
This is not Alpine.

this is how JavaScript works.

`this` inside method refers to the object that method is attached to, NOT whatever full object path you took to get there.

You could just test this with

```js
const data = {
            v1: 0,
            f1() { return this.v1; },
            f2() { return this.f1(); },
            inner: {
                v2: 0,
                f3() { return this.inner.v2; },
                f4() { return this.inner.f3(); },
            }
        }
        
data.inner.f4()
```

There is nothing Alpine could or would do about this, as this is a fundamental aspect of how JS works and would break many things.

Just imagine if Alpine rebound `this` on method calls of native objects like `array`. You'd not be able to do `data.array.push`.

Your methods/getters/setters should all be in the root level, or only refer to items on their same parent object.

Alternatively (not recommended but available if you can't avoid it) use `$data.inner.v2` instead of `this` (note: this will only work in inline expression `x-data` and not in `Alpine.data` registered components.)

**Reply by jonas-grobe** [reply]
I see. Thanks for explaining!
